beacon/lib/beacon/partition.ex

```diff
@@ -22,54 +23,68 @@ defmodule Beacon.Partition do
+    |> :ets.select([{{{group, :"$1"}}, [], [:"$1"]}])
   end
 
   @spec member_count(atom, Beacon.group()) :: non_neg_integer
-  def member_count(partition_name, group), do: :ets.lookup_element(partition_name, group, 3, 0)
+  def member_count(partition_name, group), do: :ets.lookup_element(partition_name, group, 2, 0)
 
   @spec member_counts(atom) :: %{Beacon.group() => non_neg_integer}
   def member_counts(partition_name) do
     partition_name
-    |> :ets.select([{{:"$1", :_, :"$2"}, [], [{{:"$1", :"$2"}}]}])
+    |> :ets.select([{{:"$1", :"$2"}, [], [{{:"$1", :"$2"}}]}])
     |> Map.new()
   end
 
   @spec member?(atom, Beacon.group(), pid) :: boolean
   def member?(partition_name, group, pid) do
-    case :ets.lookup_element(partition_name, group, 2, []) do
+    partition_name
+    |> Beacon.Supervisor.partition_entries_table()
+    |> :ets.lookup({group, pid})
+    |> case do
+      [{{^group, ^pid}}] -> true
       [] -> false
-      pids -> MapSet.member?(pids, pid)
     end
   end
 
   @spec groups(atom) :: [Beacon.group()]
-  def groups(partition_name), do: :ets.select(partition_name, [{{:"$1", :_, :_}, [], [:"$1"]}])
+  def groups(partition_name), do: :ets.select(partition_name, [{{:"$1", :_}, [], [:"$1"]}])
 
   @spec group_count(atom) :: non_neg_integer
   def group_count(partition_name), do: :ets.info(partition_name, :size)
 
-  @spec start_link(atom, atom) :: GenServer.on_start()
-  def start_link(scope, partition_name),
-    do: GenServer.start_link(__MODULE__, [scope, partition_name], name: partition_name)
+  @spec start_link(atom, atom, atom) :: GenServer.on_start()
+  def start_link(scope, partition_name, partition_entries_table),
+    do:
+      GenServer.start_link(__MODULE__, [scope, partition_name, partition_entries_table],
+        name: partition_name
+      )
 
   @impl true
   @spec init(any) :: {:ok, State.t()}
-  def init([scope, name]) do
-    {:ok, %State{scope: scope, name: name}, {:continue, :rebuild_monitors}}
+  def init([scope, name, entries_table]) do
+    {:ok, %State{scope: scope, name: name, entries_table: entries_table},
+     {:continue, :rebuild_monitors}}
   end
 
   @impl true
   @spec handle_continue(:rebuild_monitors, State.t()) :: {:noreply, State.t()}
   def handle_continue(:rebuild_monitors, state) do
-    monitors =
-      for {group, pids, _counter} <- :ets.tab2list(state.name), pid <- pids, into: %{} do
+    # Here we delete all counters and rebuild them based on entries table
+    :ets.delete_all_objects(state.name)
+
+    {monitors, counts} =
+      :ets.tab2list(state.entries_table)
+      |> Enum.reduce({%{}, %{}}, fn {{group, pid}}, {monitors_acc, counts_acc} ->
         ref = Process.monitor(pid, tag: {:DOWN, group})
-        {{group, pid}, ref}
-      end
+        monitors_acc = Map.put(monitors_acc, {group, pid}, ref)
+        counts_acc = Map.update(counts_acc, group, 1, &(&1 + 1))
+        {monitors_acc, counts_acc}
+      end)
+
+    :ets.insert(state.name, Enum.to_list(counts))
```

Comment:
There is a chance building this directly on ETS and using `:ets.update_counter` will be quite faster than maps for 1k+ entries. :)